[BEAM-3506] - Add a feature in JdbcIO that allows writing PCollection<Iterable<T>>

olavloite opened the pull request:
This PR adds a feature to JdbcIO that allows PCollection<Iterable<T>> to be the input of a write in addition to the current PCollection<T>. Each Iterable<T> is written to the database as one batch. This allows the user to determine the size of the batches, and also to have different sizes of batches based on the actual data.
Writing data in (large) batches has a significantly positive effect on the performance when writing to cloud databases such as Google Cloud Spanner. This example project illustrates how to use this feature in combination with Google Cloud Spanner: [https://github.com/olavloite/spanner-beam-example](https://github.com/olavloite/spanner-beam-example)

jbonofre commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:
```diff
@@ -619,6 +623,122 @@ public void teardown() throws Exception {
+    }
+
+    private static class WriteIterableFn<T> extends DoFn<Iterable<T>, Void> {
+      private static final int DEFAULT_BATCH_SIZE = 1000;
```
In the case of `WriteIteratble` I would not use batch here.

jbonofre replied:
Just to be clear: this batch size is when we have a single element. Here with a `PCollection<Iterable>`, the batch size is basically the size of the `Iterable` element.

jbonofre commented:
You should also squash to reduce the "commits noise".

jbonofre commented:
By the way, I added a method to define the batch size in #4458 . You can take a look if you want. I will try to merge pretty fast, you can rebase your branch on this one.

olavloite commented:
I refactored the code and created an abstract super class for both write operations, and thereby also removed the use of the batch size in WriteIterable.

olavloite commented:
I'll rebase my PR on your changes as soon as I can (as far as I know, I can't currently base it on your  branch as it is in your fork, so I'll have to wait until it's merged).

jkff commented:
Quick note: this is a pretty significant change; JB, please don't merge until my LGTM as well. I'll try to get to this within the next couple of days.

jkff commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:
```diff
@@ -510,69 +514,38 @@ public void teardown() throws Exception {
-      private Connection connection;
-      private PreparedStatement preparedStatement;
-      private int batchCount;
+      protected DataSource dataSource;
```
These all should be transient

jkff commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:
```diff
@@ -510,69 +514,38 @@ public void teardown() throws Exception {
+            createWriteFn(AbstractWrite<RowT, InputT> write);
 
-      private final Write<T> spec;
+    abstract static class AbstractWriteFn<RowT, InputT> extends DoFn<InputT, Void> {
```
Can this be achieved with composition rather than inheritance? (same about the enclosing transform) - composition is generally much easier to deal with.

jkff commented on sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java:
```diff
@@ -316,4 +319,68 @@ public void setParameters(KV<Integer, String> element, PreparedStatement stateme
+        }
+      }
+    } finally {
+      DatabaseTestHelper.deleteTable(dataSource, tableName);
```
Can this be in a `@Teardown` method?

jkff commented on sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java:
```diff
@@ -316,4 +319,68 @@ public void setParameters(KV<Integer, String> element, PreparedStatement stateme
+
+      pipeline.run();
+
+      try (Connection connection = dataSource.getConnection()) {
```
Can you instead use JdbcIO.read() and PAssert for this? "Write-then-read" is the recommended style for tests of IOs.

jkff commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:
```diff
@@ -635,6 +596,131 @@ public void teardown() throws Exception {
+          spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+          preparedStatement.addBatch();
+
+          batchCount++;
```
It seems batchCount is never used

jkff commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:
```diff
@@ -635,6 +596,131 @@ public void teardown() throws Exception {
+        for (T record : records) {
+          preparedStatement.clearParameters();
+          spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+          preparedStatement.addBatch();
```
What if the Iterable has too many elements for the database to handle in a single batch prepared statement?

olavloite commented:
Thanks for the review. The reason that I want to add the writeIterable method, is that it gives me more control over the actual batch size being written to the database. In my development environment, the records were always written in batches of 1 when using the original write method (even though the default batch size of 1,000 was set), which for Cloud Spanner is a really slow way of writing the data. I have not been able to find any method to control the actual batch size reliably.
Whether the default of 1,000 records per batch would be appropriate specifically for Cloud Spanner, depends on the size of the rows. Cloud Spanner has a maximum commit size of 100MB, which means that the 1,000 limit could be too low for tables with few (and small) columns, while it could be too high for tables with a large number of columns. As Jdbc is a generic standard, I guess it's hard to have one size that really fits them all, although a default of 1,000 sounds reasonable to me.

jkff commented:
It sounds like the original issue is that you were unexpectedly getting bundles of size 1 element which is indeed too small, and definitely has more performance issues than just JDBC batching. I'd love to know more about your pipeline to investigate why that would be the case (were you using the Dataflow runner or a different one? What does the pipeline look like?)

You can try to resolve that by applying the original JdbcIO.write() to just what you did here in your test + a Flatten.iterables() (in general, c.apply(JdbcIO.writeIterables()) == c.apply(Flatten.iterables()).apply(JdbcIO.write())).

olavloite commented:
I was using the DirectRunner. I'll try your suggestion to see if it makes any difference.
My Pipeline was the standard WordCount example, but writing the words to a Cloud Spanner database instead of text files. ([https://github.com/olavloite/spanner-beam-example/blob/master/src/main/java/org/apache/beam/examples/WordCount.java](https://github.com/olavloite/spanner-beam-example/blob/master/src/main/java/org/apache/beam/examples/WordCount.java))

jkff commented:
Oh I see. Direct runner indeed uses bundles of size 1 key after a grouping operation (and CountWords in that example is a grouping operation). My suggestion is likely to help - please let me know if it doesn't.

olavloite commented:
I can confirm that using the original JdbcIO.write() method on the DataflowRunner does group the writes into batches. The size of one batch is at most batchSize() elements, but sometimes also smaller. This does however solve the original problem on the DirectRunner where the batches always contain only one element. I also compared the performance of the original JdbcIO.write() method with the JdbcIO.writeIterable() method on Cloud Spanner using the DataflowRunner. I could not measure any significant difference. This means that, at least as far as I'm concerned, the functionality for writeIterable() is no longer needed.
I still updated the PR based on the review and my own test results. Feel free to use (parts of) it if you see other use cases where this functionality might be needed, or otherwise just close it.

jkff commented:
Apologies, I forgot about this PR. But yeah, we have a workaround and this issue didn't come up again, so I'm going to close. Thanks!
